master/internal/api_workspace.go

```diff
@@ -56,19 +60,45 @@ func maskStorageConfigSecrets(w *workspacev1.Workspace) error {
 	return nil
 }
 
+func validatePostWorkspaceRequest(req *apiv1.PostWorkspaceRequest) error {
+	if (req.ClusterName != nil || req.Quota != nil) &&
+		(req.NamespaceName == nil && !req.AutoCreateNamespace) {
+		return status.Errorf(codes.InvalidArgument,
+			"Must specify either an existing Kubernetes namespace or indicate that you would like a namespace to be auto-created")
+	}
+	if req.NamespaceName != nil && req.Quota != nil {
+		return status.Errorf(codes.InvalidArgument,
+			"You cannot set a quota on an already existing k8s namespace.")
+	}
+	if (req.AutoCreateNamespace || req.NamespaceName != nil) && req.ClusterName == nil {
+		return status.Errorf(codes.InvalidArgument,
+			"You must specify a cluster for the specified namespace that you would like to bind.")
+	}
+	return nil
+}
+
 func validateWorkspaceName(name string) error {
 	switch {
 	case len(name) < 1:
 		return status.Errorf(codes.InvalidArgument, "name '%s' must be at least 1 character long", name)
-	case len(name) > 80:
-		return status.Errorf(codes.InvalidArgument, "name '%s' must be at most 80 character long", name)
+	case len(name) > 53:
+		return status.Errorf(codes.InvalidArgument, "name '%s' must be at most 53 character long", name)
 	case len(strings.TrimFunc(name, unicode.IsSpace)) == 0:
 		return status.Error(codes.InvalidArgument, "name must contain at least non-whitespace letter")
 	default:
 		return nil
 	}
 }
 
+func generateNamespaceName(workspace string) (*string, error) {
+	namespace := "det-" + workspace
+	// Ensure the namespace name is <= 63 characters.
+	if len(namespace) > 63 {
+		return nil, status.Error(codes.InvalidArgument, "The namespace name must be at most 63 characters")
+	}
```

Comment:
shouldn't you still try to return _something_ here? maybe truncate the generated name to just under 63 characters if the pattern generates something too long?

Reply:
The issue with truncating is that if two people create workspace names whose generated namespace names are too long, but their first 59 characters are identical, the generated namespace names will also be identical. 
Since the `det-` prefix only adds 4 additional characters to the namespace name, I think it's a fair request to prompt the user for a more unique workspace name, as they have 59 "degrees of freedom". WDYT?